### PR TITLE
View public holiday and optional holiday cards on leave management

### LINKED ISSUE #1735: Show public holiday and optional holiday cards on leave management

### PR DESCRIPTION
**Pull request by prasanthchaduvula**
Closes #1735 

### What
- Fixed the time tracking bug if leaves are not defined for the year.
- Calculate optional holiday timeoff entries using the service.
- Updated leave balance API with optional and national holiday timeoff entries.
- View national and optional holiday cards on leave management and click on those cards to view the applied holidays respectively.

### CODE REVIEW

**Comment by nisusam** [comment]
Can you please fix the specs @prasanthchaduvula ?

File: app/javascript/src/components/LeaveManagement/Container/LeaveBlock.tsx [hunk 1]
```diff
@@ -29,7 +43,9 @@ const LeaveBlock = ({ leaveType, setSelectedLeaveType }) => {
       <div className="mt-4 flex flex-col">
         <span className="text-xs font-semibold lg:text-sm">{name}</span>
         <span className="mt-2 text-base font-semibold lg:text-2xl">
-          {`${minToHHMM(netDuration)} h (${netDays} days)`}
+          {category != "national" &&
+            `${minToHHMM(netDuration)} h (${netDays} days)`}
+          {category == "optional" && `this ${timePeriod}`}
```
Comment (nisusam) [inline app/javascript/src/components/LeaveManagement/Container/LeaveBlock.tsx:48]: Is there any other category than `national` and `optional`?

File: app/views/internal_api/v1/timeoff_entries/index.json.jbuilder [hunk 2]
```diff
@@ -10,6 +10,23 @@ json.timeoff_entries timeoff_entries do |timeoff_entry|
...
+  json.holiday_info timeoff_entry.holiday_info if timeoff_entry.holiday_info.present?
+  json.type timeoff_entry.holiday_info.present? ? "holiday" : "leave"
+end
+
+json.national_timeoff_entries national_timeoff_entries do |timeoff_entry|
+  json.id timeoff_entry.id
+  json.duration timeoff_entry.duration
+  json.leave_date CompanyDateFormattingService.new(timeoff_entry.leave_date, company: current_company).process
+  json.holiday_info timeoff_entry.holiday_info if timeoff_entry.holiday_info.present?
+  json.type timeoff_entry.holiday_info.present? ? "holiday" : "leave"
```
Comment (nisusam) [inline app/views/internal_api/v1/timeoff_entries/index.json.jbuilder:29]: Can we send `timeoff_type` as `national` or `optional`?

**Review by nisusam — APPROVED** [review]
Please fix the failing specs

File: app/javascript/src/components/LeaveManagement/Container/LeaveBlock.tsx (hunk 1 above)
Comment (prasanthchaduvula) [reply app/javascript/src/components/LeaveManagement/Container/LeaveBlock.tsx:48]: No

File: app/views/internal_api/v1/timeoff_entries/index.json.jbuilder (hunk 2 above)
Comment (prasanthchaduvula) [reply app/views/internal_api/v1/timeoff_entries/index.json.jbuilder:29]: No, as we need two different arrays on client side and if we make it into one array with timeoff_type as national or optional then we have to filter on client side.

**Pull request merged**